nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FlattenJson.java

```diff
@@ -0,0 +1,134 @@
+            .build();
+    static final Relationship REL_ORIGINAL = new Relationship.Builder()
+            .description("The original files are copied to this relationship; autoterminated by default.")
+            .autoTerminateDefault(true)
+            .name("original")
+            .build();
+
+    static final PropertyDescriptor SEPARATOR = new PropertyDescriptor.Builder()
+            .name("flatten-json-separator")
+            .displayName("Separator")
+            .defaultValue(".")
+            .description("The separator character used for joining keys. Must be a JSON-legal character.")
+            .addValidator((subject, input, context) -> {
+                boolean valid = input != null && input.length() == 1;
+                String message = !valid ? "The separator must be a single character in length." : "";
+
+                return new ValidationResult.Builder().subject(subject).input(input).valid(valid).explanation(message).build();
+            })
+            .expressionLanguageSupported(true)
+            .build();
+
+    private List<PropertyDescriptor> properties;
+    private Set<Relationship> relationships;
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        List<PropertyDescriptor> props = new ArrayList<>();
+        props.add(SEPARATOR);
+        properties = Collections.unmodifiableList(props);
+
+        Set<Relationship> rels = new HashSet<>();
+        rels.add(REL_SUCCESS);
+        rels.add(REL_FAILURE);
+        rels.add(REL_ORIGINAL);
+
+        relationships = Collections.unmodifiableSet(rels);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        FlowFile cloned = session.clone(flowFile);
+        String separator = context.getProperty(SEPARATOR).evaluateAttributeExpressions(flowFile).getValue();
+
+        try {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            session.exportTo(flowFile, bos);
+            bos.close();
+
+            String raw = new String(bos.toByteArray());
+            final String flattened = new JsonFlattener(raw)
+                    .withFlattenMode(FlattenMode.KEEP_ARRAYS)
+                    .withSeparator(separator.charAt(0))
+                    .flatten();
+
+            flowFile = session.write(flowFile, os -> os.write(flattened.getBytes()));
+
+            session.transfer(flowFile, REL_SUCCESS);
+        } catch (Exception ex) {
+            session.transfer(flowFile, REL_FAILURE);
+        } finally {
+            session.transfer(cloned, REL_ORIGINAL);
```

Comment:
All of the processors I checked that have an "original" relationship will only output the original flow file on successful transformation (i.e. not transferred to failure), and usually only when an incoming flow file can generate multiple output flow files. Are you sure you need/want an "original" relationship? Perhaps instead you can just overwrite the incoming flow file, sending the changed contents to success, or the original to failure depending on the outcome of the operation. If you do want an "original", I recommend only sending the flow file to original if successful, as the original flow file's contents will be the contents of the one on the "failure" relationship, right?